### PR TITLE
New EF.Functions.Collate

### LINKED ISSUE #8813: Support per-operation collation using EF.Functions.Collate() or equivalent

### PR DESCRIPTION
**Pull request by roji**
Closes #8813

### CODE REVIEW

File: src/EFCore.Relational/Extensions/RelationalDbFunctionsExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        ///     <para>
+        ///         The available collations and their names vary across databases, consult your database's documentation for more
+        ///         information.
+        ///     </para>
+        /// </summary>
+        /// <typeparam name="TProperty"> The type of the operand on which the collation is being specified. </typeparam>
+        /// <param name="_"> The DbFunctions instance. </param>
+        /// <param name="operand"> The operand to which to apply the collation. </param>
+        /// <param name="collation"> The name of the collation. </param>
+        public static TProperty Collate<TProperty>(
```
Comment (roji) [inline src/EFCore.Relational/Extensions/RelationalDbFunctionsExtensions.cs:28]: Generic since non-string types may be collatable (plus arbitrary CLR types mapped to database text via value converters)

File: src/EFCore.Relational/Query/ISqlExpressionFactory.cs [hunk 2]
```diff
@@ -94,6 +94,8 @@ public interface ISqlExpressionFactory
         CaseExpression Case([NotNull] SqlExpression operand, [NotNull] params CaseWhenClause[] whenClauses);
         CaseExpression Case([NotNull] IReadOnlyList<CaseWhenClause> whenClauses, [CanBeNull] SqlExpression elseResult);
 
+        CollateExpression Collate([NotNull] SqlExpression operand, [NotNull] string collation);
```
Comment (roji) [inline src/EFCore.Relational/Query/ISqlExpressionFactory.cs:97]: AFAICT COLLATE never modifies the type mapping, it just bubbles it up from its operand. So I'm not so sure that constructing CollateExpression via the factory is actually needed, did it for now for consistency. Let me know @smitpatel.

File: src/EFCore.Relational/Query/Internal/CollateTranslator.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+using System.Collections.Generic;
+using System.Reflection;
+using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore.Internal;
+using Microsoft.EntityFrameworkCore.Query.SqlExpressions;
+using Microsoft.EntityFrameworkCore.Utilities;
+
+namespace Microsoft.EntityFrameworkCore.Query.Internal
+{
+    public class CollateTranslator : IMethodCallTranslator
```
Comment (roji) [inline src/EFCore.Relational/Query/Internal/CollateTranslator.cs:13]: Number of translators is slowly creeping up again :fearful: We can merge at some point if we care.

File: src/EFCore.Relational/Query/Internal/CollateTranslator.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        private static readonly MethodInfo _methodInfo
+            = typeof(RelationalDbFunctionsExtensions).GetMethod(nameof(RelationalDbFunctionsExtensions.Collate));
+
+        private readonly ISqlExpressionFactory _sqlExpressionFactory;
+
+        public CollateTranslator([NotNull] ISqlExpressionFactory sqlExpressionFactory)
+            => _sqlExpressionFactory = sqlExpressionFactory;
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/CollateTranslator.cs:21]: We don't use expression body for ctors.

File: src/EFCore.Relational/Query/SqlExpressionFactory.cs [hunk 5]
```diff
@@ -495,6 +495,14 @@ public virtual CaseExpression Case(IReadOnlyList<CaseWhenClause> whenClauses, Sq
             return new CaseExpression(typeMappedWhenClauses, elseResult);
         }
 
+        public virtual CollateExpression Collate(SqlExpression operand, string collation)
+        {
+            Check.NotNull(operand, nameof(operand));
+            Check.NotNull(collation, nameof(collation));
+
+            return (CollateExpression)ApplyDefaultTypeMapping(new CollateExpression(operand, collation));
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressionFactory.cs:503]: How does type mapping and collation interact though?
Can you specify collation on non-string columns?
On string columns, can you still specify from outside?
e.g. `column == "a" Collate John Doe` ? Then assigning default type mapping would be wrong.

File: src/EFCore.Relational/Query/Internal/CollateTranslator.cs (hunk 4 above)
Comment (roji) [reply src/EFCore.Relational/Query/Internal/CollateTranslator.cs:21]: I sense a fight :) (just kidding)

Any reason to make an exception for constructors as opposed to other methods (or properties)? Seems quite inconsistent to me...

File: src/EFCore.Relational/Query/SqlExpressionFactory.cs (hunk 5 above)
Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs:503]: > How does type mapping and collation interact though?
>  Can you specify collation on non-string columns?

Collation support is typically restricted to textual types. However, that list isn't necessarily well-known, and conceptually other text-like types may also support collation (json/xml/full text search, although I haven't done the research here yet). So I preferred to not restrict anything here - since the user is very explicitly using EF.Functions.Collate, my logic was to translate this and let the database error if not supported.

We could add a flag on RelationalTypeMapping to indicate whether it's collatable, but as it would only serve to throwing earlier it doesn't seem very valuable... What are your thoughts?

> On string columns, can you still specify from outside?
> e.g. column == "a" Collate John Doe ? Then assigning default type mapping would be wrong.

What happens here is subject to database collation precedence rules. For example, in your expression "a" has an explicit collation, so that takes precedence over any collation configured on column. This seems to be pretty cross-database (but again, haven't researched exhaustively). Do you see a potential issue here?

File: src/EFCore.Relational/Query/Internal/CollateTranslator.cs (hunk 4 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/Internal/CollateTranslator.cs:21]: I don't know. I just mentioned what is happening in rest of codebase. I don't have any preference. We can change the rule also.

File: src/EFCore.Relational/Query/SqlExpressionFactory.cs (hunk 5 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs:503]: For first one, I was thinking, if it is only on string type then we can check TypeMapping.StoreType to throw early. Since TypeMapping would be available. Was trying to think of more understandable exception rather than just invalid SQL which user need to parse to understand the error. But if it is too difficult then we can just let Invalid SQL like we do in other places.

For second one, My question is about TypeMapping rather than collation precedence rules. What typemapping should be used to print out "a", should it be unicode or non-unicode? Does collation affect that aspect?

File: src/EFCore.Relational/Query/Internal/CollateTranslator.cs (hunk 4 above)
Comment (roji) [reply src/EFCore.Relational/Query/Internal/CollateTranslator.cs:21]: OK. We can see if anyone has any objections in the design meeting.

File: src/EFCore.Relational/Query/SqlExpressionFactory.cs (hunk 5 above)
Comment (lauxjpn) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs:503]: > For second one, My question is about TypeMapping rather than collation precedence rules. What typemapping should be used to print out "a", should it be unicode or non-unicode? Does collation affect that aspect?

For MySQL, explicitly using a collation can change the underlying char set for that expression, if the collation used is from a different char set (every collation depends on a single char set in MySQL).

Internally in Pomelo, we do not do anything with the Unicode/non-Unicode `StringTypeMapping` flag though and just ignore it. We only care about, whether a char set and/or collation was set or not.

Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs:503]: > For first one, I was thinking, if it is only on string type then we can check TypeMapping.StoreType to throw early. Since TypeMapping would be available. Was trying to think of more understandable exception rather than just invalid SQL which user need to parse to understand the error. But if it is too difficult then we can just let Invalid SQL like we do in other places.

OK. A check on the CLR type seems problematic, because value converters can be used to convert a non-string CLR type to a database text type, and users may want to apply collation to those. Another objection is text-like database types that don't map to CLR string (json/xml/full text), so we should be careful about restricting this to CLR string.

A check on the store type could work better, but then the provider would have to tell us which type mappings are collatable, and which aren't (because the list is provider-specific). We can add IsCollatable to RelationalTypeMapping, but if we do it only in order to raise an exception it seems like low-value to me...

(BTW I actually think database errors aren't necessarily a bad thing or something we should hide/improve on, but of course it's case by case)

> For second one, My question is about TypeMapping rather than collation precedence rules. What typemapping should be used to print out "a", should it be unicode or non-unicode? Does collation affect that aspect?

tl;dr "a" should have the same type mapping as it would have had if COLLATE wasn't there (which seems to be varchar on SQL Server).

More in-depth... unicode/non-unicode as a *type mapping* thing seems to be SQL Server-specific (this is why at some point I didn't like that RelationlTypeMapping has IsUnicode on it...).

* SQL Server: varchar vs. nvarchar types as we know
* [PostgreSQL](https://www.postgresql.org/docs/current/multibyte.html): set character set is determined at the database level only (e.g. `CREATE DATABASE korean WITH ENCODING 'EUC_KR'`).
* [MySQL](https://dev.mysql.com/doc/refman/8.0/en/string-type-syntax.html) (see also [this on unicode](https://dev.mysql.com/doc/refman/8.0/en/charset-unicode.html)): character set can be set per-column.
* Sqlite: seems to be always Unicode, though I'm not 100% sure.

I did a bit of testing to see what the COLLATE expression does to type mappings on SQL Server:

```sql
CREATE TABLE data (
    c1 varchar(255),
    c2 nvarchar(255)
);
INSERT INTO data (c1, c2) VALUES ('foo', 'foo');

SELECT SQL_VARIANT_PROPERTY(c1, 'BaseType'), SQL_VARIANT_PROPERTY(c1 COLLATE Latin1_General_CS_AS, 'BaseType') FROM data;
-- returns varchar, varchar
SELECT SQL_VARIANT_PROPERTY(c2 , 'BaseType'), SQL_VARIANT_PROPERTY(c2 COLLATE Latin1_General_CS_AS, 'BaseType') FROM data;
-- returns nvarchar, nvarchar

SELECT SQL_VARIANT_PROPERTY(c1, 'MaxLength'), SQL_VARIANT_PROPERTY(c1 COLLATE Latin1_General_CS_AS, 'MaxLength') FROM data;
-- returns 255, 255
SELECT SQL_VARIANT_PROPERTY(c2 , 'MaxLength'), SQL_VARIANT_PROPERTY(c2 COLLATE Latin1_General_CS_AS, 'MaxLength') FROM data;
-- returns 510, 510
```

In other words, the expression seems to be completely pass-through as far as type mapping is concerned - it doesn't change anything, just applies a collation to be used for sorting/equality. Does this seem exhaustive enough, do you think there are further open questions?

(BTW there's a connection between character sets and collations - the set of collations you can use is determined by the character set - but that doesn't seem to be something we should care about).

Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs:503]: @lauxjpn

> For MySQL, explicitly using a collation can change the underlying char set for that expression, if the collation used is from a different char set (every collation depends on a single char set in MySQL).

Am curious, can you give an example of how that works? In general, I'm not used to thinking of an *expression* having a char set (as opposed to a column or database).

> Internally in Pomelo, we do not do anything with the Unicode/non-Unicode StringTypeMapping flag though and just ignore it

Same in Npgsql.

Comment (lauxjpn) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs:503]: > Am curious, can you give an example of how that works? In general, I'm not used to thinking of an expression having a char set (as opposed to a column or database).

> ~~For MySQL, explicitly using a collation *can* change the underlying char set for that expression, if the collation used is from a different char set (every collation depends on a single char set in MySQL).~~

Scrap that. String literal *have* a char set in MySQL, but it is always the one specified for the connection and an explicit collation must be compatible with it:

(From [10.3.6 Character String Literal Character Set and Collation](https://dev.mysql.com/doc/refman/8.0/en/charset-literal.html))

>`[_charset_name]'string' [COLLATE collation_name]`
>
>[...]
> * If `_charset_name` is not specified but `COLLATE collation_name` is specified, the [connection default character set](https://dev.mysql.com/doc/refman/8.0/en/charset-connection.html) given by the `character_set_connection` system variable and collation `collation_name` are used. `collation_name` must be a permitted collation for the connection default character set.

Comment (lauxjpn) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs:503]: > ~~For MySQL, explicitly using a collation *can* change the underlying char set for that expression, if the collation used is from a different char set (every collation depends on a single char set in MySQL).~~

Scrap that. String literals *have* a char set in MySQL, but it is always the one specified for the connection and an explicit collation must be compatible with it:

(From [10.3.6 Character String Literal Character Set and Collation](https://dev.mysql.com/doc/refman/8.0/en/charset-literal.html))

>`[_charset_name]'string' [COLLATE collation_name]`
>
>[...]
> * If `_charset_name` is not specified but `COLLATE collation_name` is specified, the [connection default character set](https://dev.mysql.com/doc/refman/8.0/en/charset-connection.html) given by the `character_set_connection` system variable and collation `collation_name` are used. `collation_name` must be a permitted collation for the connection default character set.

> Am curious, can you give an example of how that works? In general, I'm not used to thinking of an expression having a char set (as opposed to a column or database).

It is possible to explicitly change the char set of the string literal, but not just by the collation, but by also specifying a *char set introducer*:

> * If both `_charset_name` and `COLLATE collation_name` are specified, character set `charset_name` and collation `collation_name` are used. `collation_name` must be a permitted collation for `charset_name`.

Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs:503]: Great info, thanks @lauxjpn.

Comment (smitpatel) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs:503]: Then the code is incorrect here.

Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs]: @smitpatel you're right, thanks - corrected to just always bubble up the type mapping. Though is there any value in having this method on SqlExpressionFactory, rather than just constructing CollateExpression directly?

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        }
+
+        public virtual SqlExpression Operand { get; }
+        public virtual string Collation { get; }
+
+        protected override Expression VisitChildren(ExpressionVisitor visitor)
+            => Update((SqlExpression)visitor.Visit(Operand));
+
+        public virtual CollateExpression Update([NotNull] SqlExpression operand)
+            => operand == Operand
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs:34]: Invert condition. Add Check.NotNull

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+using Microsoft.EntityFrameworkCore.Utilities;
+
+namespace Microsoft.EntityFrameworkCore.Query.SqlExpressions
+{
+    public class CollateExpression : SqlExpression
+    {
+        public CollateExpression(
+            [NotNull] SqlExpression operand,
+            [NotNull] string collation,
+            [CanBeNull] RelationalTypeMapping typeMapping = null)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs:17]: If type mapping can never change and remove type mapping argument here. And always pass in from operand to base.

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        public override bool Equals(object obj)
+            => obj != null
+                && (ReferenceEquals(this, obj)
+                    || obj is CollateExpression likeExpression
+                    && Equals(likeExpression));
+
+        private bool Equals(CollateExpression collateExpression)
+            => base.Equals(collateExpression)
+                && Operand.Equals(collateExpression.Operand)
+                && Collation == collateExpression.Collation;
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs:57]: Use string.Equals with appropriate comparison option.

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+            expressionPrinter.Visit(Operand);
+            expressionPrinter
+                .Append(" COLLATE ")
+                .Append(Collation);
+        }
+
+        public override bool Equals(object obj)
+            => obj != null
+                && (ReferenceEquals(this, obj)
+                    || obj is CollateExpression likeExpression
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs:51]: variable name.

File: src/EFCore.Relational/Query/SqlExpressionFactory.cs [hunk 10]
```diff
@@ -495,6 +495,14 @@ public virtual CaseExpression Case(IReadOnlyList<CaseWhenClause> whenClauses, Sq
             return new CaseExpression(typeMappedWhenClauses, elseResult);
         }
 
+        public virtual CollateExpression Collate(SqlExpression operand, string collation)
+        {
+            Check.NotNull(operand, nameof(operand));
+            Check.NotNull(collation, nameof(collation));
+
+            return new CollateExpression(operand, collation);
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressionFactory.cs:503]: If there cannot be anyway CollateExpression can have different type mapping than operand then you can remove this method and initialize using new syntax.

File: src/EFCore.Relational/Query/SqlExpressionFactory.cs [hunk 11]
```diff
@@ -495,6 +495,14 @@ public virtual CaseExpression Case(IReadOnlyList<CaseWhenClause> whenClauses, Sq
             return new CaseExpression(typeMappedWhenClauses, elseResult);
         }
 
+        public virtual CollateExpression Collate(SqlExpression operand, string collation)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressionFactory.cs:498]: Bubbling up still does not work. You need to also add code how to apply typeMapping from outside to CollateExpression.

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs (hunk 7 above)
Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs:17]: I think this is the case, but the idea is to leave an escape hatch in case I've missed something, or some other unknown database has peculiar behavior. Any objections to just leaving it?

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs (hunk 8 above)
Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs:57]: I think we may have discussed this, but equality operator for strings is the same as `string.Equals(s, StringComparison.Ordinals)`, but the latter is needlessly verbose... Will change but I think we should reconsider.

File: src/EFCore.Relational/Query/SqlExpressionFactory.cs (hunk 10 above)
Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs:503]: I agree, but assuming we leave the escape hatch as per https://github.com/dotnet/efcore/pull/20603#pullrequestreview-393031809, would you want to leave this?

File: src/EFCore.Relational/Query/SqlExpressionFactory.cs (hunk 11 above)
Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs:498]: Thanks, added support in ApplyTypeMapping.

**Comment by roji** [comment]
@smitpatel addressed review comment.

File: src/EFCore.Relational/Query/SqlExpressionFactory.cs [hunk 12]
```diff
@@ -58,6 +58,7 @@ public virtual SqlExpression ApplyTypeMapping(SqlExpression sqlExpression, Relat
             return sqlExpression switch
             {
                 CaseExpression e => ApplyTypeMappingOnCase(e, typeMapping),
+                CollateExpression e => e.ApplyTypeMapping(typeMapping),
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressionFactory.cs:61]: That is not how collate is supposed to apply type mapping. 🤦

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs (hunk 7 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs:17]: If you think we need escape hatch then we need more research on how collation affects type mapping since it changes a lot of things in type mapping inference.
I would be on the side of not adding escape hatch. If your type mapping differs then this CollateExpression is not for you. Add your custom one which also deals with inferring/propagating typemapping.

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+            Check.NotNull(collation, nameof(collation));
+
+            Operand = operand;
+            Collation = collation;
+        }
+
+        public virtual SqlExpression Operand { get; }
+        public virtual string Collation { get; }
+
+        public SqlExpression ApplyTypeMapping([CanBeNull] RelationalTypeMapping typeMapping)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs:30]: Remove

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    public class CollateExpression : SqlExpression
+    {
+        public CollateExpression(
+            [NotNull] SqlExpression operand,
+            [NotNull] string collation,
+            [CanBeNull] RelationalTypeMapping typeMapping = null)
+            : base(operand.Type, typeMapping ?? operand.TypeMapping)
+        {
+            Check.NotNull(operand, nameof(operand));
+            Check.NotNull(collation, nameof(collation));
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs:21]: Is empty string as collation valid? If not then Check.NotEmpty

File: src/EFCore.Relational/Query/SqlExpressionFactory.cs (hunk 12 above)
Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs:61]: Sorry about that, am going a bit too fast today.

**Comment by roji** [comment]
Sorry about the previous nonsense, hopefully I got it right this time.

**Comment by smitpatel** [comment]
@roji be like

File: src/EFCore.Relational/Query/NullabilityBasedSqlProcessingExpressionVisitor.cs [hunk 15]
```diff
@@ -169,6 +169,14 @@ protected override Expression VisitCase(CaseExpression caseExpression)
             return caseExpression.Update(operand, whenClauses, elseResult);
         }
 
+        protected override Expression VisitCollate(CollateExpression collateExpresion)
+        {
+            Check.NotNull(collateExpresion, nameof(collateExpresion));
+
+            return collateExpresion.Update(
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/NullabilityBasedSqlProcessingExpressionVisitor.cs:176]: @maumar to verify

File: src/EFCore.Relational/Query/SqlExpressionFactory.cs (hunk 10 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs:510]: We can remove the method.

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs (hunk 7 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs:17]: #unresolved

File: test/EFCore.Relational.Specification.Tests/Query/RelationalNorthwindDbFunctionsQueryTestBase.cs [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+        [MemberData(nameof(IsAsyncData))]
+        public virtual Task Collate_case_sensitive(bool async)
+            => AssertCount(
+                async,
+                ss => ss.Set<Customer>(),
+                ss => ss.Set<Customer>(),
+                c => EF.Functions.Collate(c.ContactName, CaseSensitiveCollation) == "maria anders",
+                c => c.ContactName.Equals("maria anders", StringComparison.Ordinal));
+
+        protected virtual string CaseInsensitiveCollation
```
Comment (smitpatel) [inline test/EFCore.Relational.Specification.Tests/Query/RelationalNorthwindDbFunctionsQueryTestBase.cs:40]: Just make it abstract

**Comment by roji** [comment]
Made abstract, let me know if you have any more comments.

Today @roji be more like:

image

**Comment by ajcvickers** [comment]
@roji @smitpatel I haven't read through the comments here but I noticed there are a lot! Is there anything in particular I should take a look at or needs resolving?

**Comment by roji** [comment]
@ajcvickers I don't think so, from my side everything is fine - just waiting for approval.

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs (hunk 7 above)
Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs:17]: Makes sense, am removing the parameter.

Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs:17]: > If you think we need escape hatch then we need more research on how collation affects type mapping since it changes a lot of things in type mapping inference.

The problem is that I'm not aware of a situation that requires this escape hatch, so difficult to research how collation would interact with type mapping in that case :)

This escape hatch could still make sense, since the provider could still provide their type mapping logic in their own SqlExpressionFactory. But better to not deal with theoretical situations, we can always revisit if/when a need comes up.

**Comment by roji** [comment]
@smitpatel new version up.

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs (hunk 7 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs]: My problem with escape hatch is, there is a lot more complicated code a provider has to write to deal with type mappings rather than just passing one for CollateExpression. At that point adding a custom CollateExpression to accommodate typeMapping is pretty minor point. On the other side, it is guaranteed that things won't work by just providing ability add a type mapping to CollateExpression. At the least provider needs to address, how operand sends typeMapping outside and how outside typeMapping would flow to operand.

**Comment by roji** [comment]
Rebased and made collation abstract in migrations tests too ( #20646)

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs (hunk 7 above)
Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs]: Makes sense @smitpatel, thanks.

File: src/EFCore.Relational/Query/Internal/CollateTranslator.cs (hunk 4 above)
Comment (ajcvickers) [reply src/EFCore.Relational/Query/Internal/CollateTranslator.cs:21]: I don't think we ever made an exception here as such. Expression-bodied constructors came in a later version that expression-bodied methods and we just didn't convert to them or start using them much, but I don't remember anyone having any objections to them.

Comment (smitpatel) [reply src/EFCore.Relational/Query/Internal/CollateTranslator.cs:21]: Missing Check.NotNull.

Comment (roji) [reply src/EFCore.Relational/Query/Internal/CollateTranslator.cs:21]: @smitpatel elegant way to avoid the expression-bodied constructor :)

Comment (roji) [reply src/EFCore.Relational/Query/Internal/CollateTranslator.cs:21]: But I'm do even better by eliminator it altogether (no need for SqlExpressionFactory...)

File: src/EFCore.Relational/Query/NullabilityBasedSqlProcessingExpressionVisitor.cs (hunk 15 above)
Comment (roji) [reply src/EFCore.Relational/Query/NullabilityBasedSqlProcessingExpressionVisitor.cs:176]: @maumar @smitpatel are we waiting for anything in particular here? Would be good to merge this for preview4.

Comment (smitpatel) [reply src/EFCore.Relational/Query/NullabilityBasedSqlProcessingExpressionVisitor.cs:176]: Waiting for this.

Comment (maumar) [reply src/EFCore.Relational/Query/NullabilityBasedSqlProcessingExpressionVisitor.cs:176]: this is correct - signed off

File: src/EFCore.Relational/Query/Internal/CollateTranslator.cs [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+using System.Collections.Generic;
+using System.Reflection;
+using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore.Internal;
+using Microsoft.EntityFrameworkCore.Query.SqlExpressions;
+using Microsoft.EntityFrameworkCore.Utilities;
+
+namespace Microsoft.EntityFrameworkCore.Query.Internal
+{
+    public class CollateTranslator : IMethodCallTranslator
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/CollateTranslator.cs:13]: Can you add internal docs.

File: src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+            Operand = operand;
+            Collation = collation;
+        }
+
+        public virtual SqlExpression Operand { get; }
+        public virtual string Collation { get; }
+
+        protected override Expression VisitChildren(ExpressionVisitor visitor)
+            => Update((SqlExpression)visitor.Visit(Operand));
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressions/CollateExpression.cs:30]: Check.NotNull on visitor

File: test/EFCore.Relational.Specification.Tests/Query/RelationalNorthwindDbFunctionsQueryTestBase.cs [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+                c => c.ContactName.Equals("maria anders", StringComparison.OrdinalIgnoreCase));
+
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual Task Collate_case_sensitive(bool async)
+            => AssertCount(
+                async,
+                ss => ss.Set<Customer>(),
+                ss => ss.Set<Customer>(),
+                c => EF.Functions.Collate(c.ContactName, CaseSensitiveCollation) == "maria anders",
```
Comment (smitpatel) [inline test/EFCore.Relational.Specification.Tests/Query/RelationalNorthwindDbFunctionsQueryTestBase.cs:37]: What is the effect of using EF.Functions.Collate on a string column and projecting it out?

File: src/EFCore.Relational/Query/Internal/CollateTranslator.cs (hunk 17 above)
Comment (roji) [reply src/EFCore.Relational/Query/Internal/CollateTranslator.cs:13]: These seem to be missing from all translators, will submit a separate PR for this.

File: test/EFCore.Relational.Specification.Tests/Query/RelationalNorthwindDbFunctionsQueryTestBase.cs (hunk 19 above)
Comment (roji) [reply test/EFCore.Relational.Specification.Tests/Query/RelationalNorthwindDbFunctionsQueryTestBase.cs:37]: Assuming I understand your question: the collation aspect bubbles up across projections - does this look good?

SQL Server:

```sql
DROP TABLE data;
CREATE TABLE data (
    id INT IDENTITY,
    cs VARCHAR(MAX) COLLATE Latin1_General_CS_AS
);
INSERT INTO data (cs) VALUES ('foo');

SELECT * FROM data WHERE cs = 'FOO'; -- no results because column is case-sensitive

SELECT * FROM (
    SELECT cs COLLATE Latin1_General_CI_AS AS cs
    FROM data
) AS t
WHERE t.cs = 'FOO'; -- returns row because cs projects out with the explicitly-applied collation
```

PostgreSQL:

```sql
CREATE COLLATION cs_collation (LC_COLLATE = 'en-u-ks-primary',
    LC_CTYPE = 'en-u-ks-primary',
    PROVIDER = icu,
    DETERMINISTIC = False
);

DROP TABLE IF EXISTS data;
CREATE TABLE data (
    id INT GENERATED ALWAYS AS IDENTITY,
    cs TEXT COLLATE "POSIX"
);
INSERT INTO data (cs) VALUES ('foo');

SELECT * FROM data WHERE cs = 'FOO'; -- no results because column is case-sensitive

SELECT * FROM (
    SELECT cs COLLATE cs_collation AS no_longer_cs
    FROM data
) AS t
WHERE t.no_longer_cs = 'FOO'; -- returns row because cs projects out with the explicitly-applied collation
```

**Comment by roji** [comment]
@smitpatel @ajcvickers what's your opinion on merging this into the preview4 branch? It would be good to deliver this with the other collation functionality (column/database migration support)

File: src/EFCore.Relational/Query/Internal/CollateTranslator.cs (hunk 17 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/Internal/CollateTranslator.cs:13]: Ignore. I already did this as part of query docs work. I am trying to make sure we have docs for all new stuff.

File: test/EFCore.Relational.Specification.Tests/Query/RelationalNorthwindDbFunctionsQueryTestBase.cs (hunk 19 above)
Comment (smitpatel) [reply test/EFCore.Relational.Specification.Tests/Query/RelationalNorthwindDbFunctionsQueryTestBase.cs:37]: Nooo. I wanted a test where you are using EF.Collate in projection and if it generates any difference in reading results. (which it should not.)

File: src/EFCore.Relational/Query/Internal/CollateTranslator.cs (hunk 17 above)
Comment (roji) [reply src/EFCore.Relational/Query/Internal/CollateTranslator.cs:13]: OK thanks!

File: test/EFCore.Relational.Specification.Tests/Query/RelationalNorthwindDbFunctionsQueryTestBase.cs (hunk 19 above)
Comment (roji) [reply test/EFCore.Relational.Specification.Tests/Query/RelationalNorthwindDbFunctionsQueryTestBase.cs:37]: Ohh you mean top-most projection (i.e. on a column that will be read by the client)? If so I'll check that, no problem (did you mean you want actual test code for this?)

Comment (roji) [reply test/EFCore.Relational.Specification.Tests/Query/RelationalNorthwindDbFunctionsQueryTestBase.cs:37]: Do you mean the following (which passes on both providers):

```c#
[ConditionalTheory]
[MemberData(nameof(IsAsyncData))]
public virtual Task Collate_column_projection(bool async)
    => AssertQuery(
        async,
        ss => ss.Set<Customer>().Select(c => EF.Functions.Collate(c.ContactName, CaseInsensitiveCollation)),
        ss => ss.Set<Customer>().Select(c => c.ContactName));
```

If so do you want this in the test suite?

**Comment by ajcvickers** [comment]
@roji I think it's probably okay just to wait.
